include/exadg/acoustic_conservation_equations/spatial_discretization/operators/grad_div_operator.h

```diff
@@ -0,0 +1,537 @@
+  face_loop(dealii::MatrixFree<dim, Number> const & matrix_free_in,
+            BlockVectorType &                       dst,
+            BlockVectorType const &                 src,
+            Range const &                           face_range) const
+  {
+    // precompute numbers that are needed in kernels.
+    Number const rhocc   = (Number)(data.density * data.speed_of_sound * data.speed_of_sound);
+    Number const rho_inv = (Number)(1.0 / data.density);
+    Number const tau     = (Number)(0.5 * data.speed_of_sound * data.density);
+    Number const gamma   = (Number)(0.5 / (data.speed_of_sound * data.density));
+
+    FaceIntegratorP pressure_m(matrix_free_in, true, data.dof_index_pressure, data.quad_index);
+    FaceIntegratorP pressure_p(matrix_free_in, false, data.dof_index_pressure, data.quad_index);
+
+    FaceIntegratorU velocity_m(matrix_free_in, true, data.dof_index_velocity, data.quad_index);
+    FaceIntegratorU velocity_p(matrix_free_in, false, data.dof_index_velocity, data.quad_index);
+
+    for(unsigned int face = face_range.first; face < face_range.second; face++)
+    {
+      pressure_m.reinit(face);
+      pressure_m.gather_evaluate(src.block(0), integrator_flags_p.face_evaluate);
+      pressure_p.reinit(face);
+      pressure_p.gather_evaluate(src.block(0), integrator_flags_p.face_evaluate);
+
+      velocity_m.reinit(face);
+      velocity_m.gather_evaluate(src.block(1), integrator_flags_u.face_evaluate);
+      velocity_p.reinit(face);
+      velocity_p.gather_evaluate(src.block(1), integrator_flags_u.face_evaluate);
+
+      do_face_integral<formulation, true>(
+        pressure_m, pressure_p, velocity_m, velocity_p, rho_inv, tau, rhocc, gamma);
+
+      pressure_m.integrate_scatter(integrator_flags_p.face_integrate, dst.block(0));
+      pressure_p.integrate_scatter(integrator_flags_p.face_integrate, dst.block(0));
+
+      velocity_m.integrate_scatter(integrator_flags_u.face_integrate, dst.block(1));
+      velocity_p.integrate_scatter(integrator_flags_u.face_integrate, dst.block(1));
+    }
+  }
+
+
+  template<FormulationGradientDivergenceTerms formulation>
+  void
+  boundary_face_loop(dealii::MatrixFree<dim, Number> const & matrix_free_in,
+                     BlockVectorType &                       dst,
+                     BlockVectorType const &                 src,
+                     Range const &                           face_range) const
+  {
+    // precompute numbers that are needed in kernels.
+    Number const rhocc   = (Number)(data.density * data.speed_of_sound * data.speed_of_sound);
+    Number const rho_inv = (Number)(1.0 / data.density);
+    Number const tau     = (Number)(0.5 * data.speed_of_sound * data.density);
+    Number const gamma   = (Number)(0.5 / (data.speed_of_sound * data.density));
+
+    FaceIntegratorP pressure_m(matrix_free_in, true, data.dof_index_pressure, data.quad_index);
+    BoundaryFaceIntegratorP<dim, Number> pressure_p(pressure_m, *data.bc->pressure);
```

Comment:
This seems to be the only place where the new type is used? From a general software development perspective, it is then however not clear to me why the new class `BoundaryFaceIntegratorP` is introduced at all. The interface of `do_face_integral()` could also be designed as passing the result of `ExpertiorFaceIntegratorP::get_value()` directly to `do_face_integral()`. Such a design would be clearer since it does not introduce implicit dependencies between template parameters.

Reply:
see https://github.com/exadg/exadg/pull/587/files#r1377402916